numpy/ma/core.py

```diff
@@ -4522,6 +4522,22 @@ def trace(self, offset=0, axis1=0, axis2=1, dtype=None, out=None):
             return D.astype(dtype).filled(0).sum(axis=None, out=out)
     trace.__doc__ = ndarray.trace.__doc__
 
+    def dot(self, other, out=None):
+        am = ~getmaskarray(self)
+        bm = ~getmaskarray(other)
+        if out is None:
+            d = np.dot(filled(self, 0), filled(other, 0))
+            m = ~np.dot(am, bm)
+            return masked_array(d, mask=m)
+        d = self.filled(0).dot(other.filled(0), out)
+        if out.mask.shape != d.shape:
```

Comment:
I think we need to test if out is a maskedarray (other functions do this):

            if isinstance(out, MaskedArray):
               ...